doc/developer/design/20230110_window_functions.md

```diff
@@ -0,0 +1,331 @@
+    - `full_ranges`, `zero_ranges`, `used_ranges` stay the same.
+    - `init_states` won’t start at position 0, but at the lower end of the intervals in `queries`
+    - The iteration at the end will be mostly the same.
+
+#### 3. FIRST_VALUE / LAST_VALUE / NTH_VALUE
+
+For example: For each city, compute how many times it is smaller than the biggest city in the same state:
+```sql
+SELECT state, name, pop,
+       CAST(FIRST_VALUE(pop)
+              OVER (PARTITION BY state ORDER BY pop DESC)
+            AS float
+       ) / pop
+FROM cities;
+```
+
+These also operate based on a **frame**, similarly to window aggregations. (The above example query doesn't specify a frame, therefore it uses the default frame: from the beginning of the partition to the current row) They can be similarly implemented to window aggregations, i.e., we could “sum” up the relevant interval (that is not necessarily a prefix) with an appropriate sum function.
+
+Alternatively, we could make these a bit faster (except for NTH_VALUE) if we just find the index of the relevant end of the interval (i.e., left end for FIRST_VALUE), and then self-join.
+
+(And there are some special cases when we can transform away the window function usage: FIRST_VALUE with UNBOUNDED PRECEDING and LAST_VALUE with UNBOUNDED FOLLOWING should be transformed to just a (non-windowed) grouped aggregation + self-join instead of prefix sum trickery. Also, similarly for the case when there is no ORDER BY.)
+
+----------------------
+
+### Duplicate indexes
+
+There might be rows which agree on both the PARTITION BY key and the ORDER BY key (the index). Such duplicate indexes are not handled by Prefix Sum. To eliminate these duplicates, we will number the elements inside each group with 0, 1, 2, …, and this will be an additional component of the prefix sum indexes.
+
+For this to perform well, we are assuming that groups are small.
+This is not an unreasonable assumption, because a group is identified here by a value of the PARTITION BY expression + a value of the ORDER BY expression.
+Also note that if there is no ORDER BY, then groups might be large, but in this case we don’t employ Prefix Sum, but we transform away the window functions to grouped aggregation + self-join (as noted above).
+
+Alternatively, we could change Prefix Sum so that it correctly handles duplicate indexes.
+
+### Parallelism
+
+DD's Prefix Sum should be data-parallel even inside a window partition. (It’s similar to [a Fenwick tree](https://en.wikipedia.org/wiki/Fenwick_tree), with sums maintained over power-of-2 sized intervals, from which you can compute a prefix sum by putting together LogN intervals.) TODO: But I wasn't able to actually observe a speedup in a simple test when adding cores, so we should investigate what’s going on with parallelization. There was probably just some technical issue, because all operations in the Prefix Sum implementation look parallelizable.
```

Comment:
This sounds like you don't want to make it part of the design, but I strongly suggest that any implementation includes a performance evaluation, and that we outline expected performance here.

Reply:
Ok, yes, I added much more details about performance. Will add more when I have a prototype.